test/Microsoft.TestPlatform.AcceptanceTests/Extension/TestDataSource.cs

```diff
@@ -0,0 +1,71 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Reflection;
+
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+namespace Microsoft.TestPlatform.AcceptanceTests;
+
+[AttributeUsage(AttributeTargets.Method)]
+public abstract class TestDataSource<T1> : Attribute, ITestDataSource where T1 : notnull
+{
+    private readonly List<object[]> _data = new();
+
+    public abstract void CreateData(MethodInfo methodInfo);
+
+    public void AddData(T1 value1)
+    {
+        _data.Add(new object[] { value1 });
+    }
+
+    public virtual string GetDisplayName(MethodInfo methodInfo, T1 value1)
+    {
+        return $"{methodInfo.Name} ({value1})";
+    }
+
+    IEnumerable<object[]> ITestDataSource.GetData(MethodInfo methodInfo)
+    {
+        CreateData(methodInfo);
+        return _data;
```

Comment:
Any reason not to have `CreateData` returning a collection instead of being impure? Also, it doesn't seem we are doing something with the field so maybe there is no point in holding the state (or maybe you are thinking about some later change you want to introduce).

Reply:
Yes, it is an abstract class that adds typed interface to the collection of objects we need to produce. CreateData is implemented by the inheriting class, hence why it is abstract as well. And that class then calls AddData which have the correctly typed parameters. There are probably more elegant and generic ways to do this, e.g. similar to TheoryData on xunit, but then the data source is still required to implement IDataSource and connect those two patterns together. So instead I implement a single different pattern. 

https://andrewlock.net/creating-strongly-typed-xunit-theory-test-data-with-theorydata/

I could also produce a list of tuples, but that seemed even less clear.